File: cli/python/src/memgo_cli/cli/registrations/delete.py
"""命令 delete 的参数声明与注册。"""

from __future__ import annotations

import typer
from rich.console import Console

from memgo_cli.application.entity_ids import _resolve_ids
from memgo_cli.cli.context import _fire_telemetry, _get_backend, _get_backend_and_config
from memgo_cli.output.branding import print_error

console = Console()
err_console = Console(stderr=True)


def delete(
    memory_id: str | None = typer.Argument(
        None, help="Memory ID to delete (omit when using --all or --entity)."
    ),
    all_: bool = typer.Option(False, "--all", help="Delete all memories matching scope filters."),
    entity: bool = typer.Option(
        False, "--entity", help="Delete the entity itself and all its memories (cascade)."
    ),
    project: bool = typer.Option(
        False, "--project", help="With --all: delete ALL memories project-wide."
    ),
    dry_run: bool = typer.Option(
        False, "--dry-run", help="Show what would be deleted without deleting."
    ),
    force: bool = typer.Option(False, "--force", help="Skip confirmation."),
    delete_linked: bool = typer.Option(
        False, "--delete-linked", help="Also delete memories linked to this memory."
    ),
    user_id: str | None = typer.Option(
        None, "--user-id", "-u", help="Scope to user.", rich_help_panel="Scope"
    ),
    agent_id: str | None = typer.Option(
        None, "--agent-id", help="Scope to agent.", rich_help_panel="Scope"
    ),
    app_id: str | None = typer.Option(
        None, "--app-id", help="Scope to app.", rich_help_panel="Scope"
    ),
    run_id: str | None = typer.Option(
        None, "--run-id", help="Scope to run.", rich_help_panel="Scope"
    ),
    output: str = typer.Option(
        "text", "--output", "-o", help="Output: text, json, quiet.", rich_help_panel="Output"
    ),
    api_key: str | None = typer.Option(
        None,
        "--api-key",
        help="Override API key.",
        envvar="MEMGO_API_KEY",
        rich_help_panel="Connection",
    ),
    base_url: str | None = typer.Option(
        None, "--base-url", help="Override API base URL.", rich_help_panel="Connection"
    ),
) -> None:
    """校验删除选项互斥关系并分发单条、范围或实体删除。"""
    # 校验选项互斥关系
    modes = sum([memory_id is not None, all_, entity])
    if modes > 1:
        print_error(
            err_console,
            "Only one of memory ID, --all, or --entity may be used at a time.",
            hint=None,
        )
        raise typer.Exit(1)
    if modes == 0:
        print_error(
            err_console,
            "Provide a memory ID, --all, or --entity.",
            hint="Run 'memgo delete --help' for usage.",
        )
        raise typer.Exit(1)

    # 按删除目标分发
    if memory_id is not None:
        _fire_telemetry("delete", {"delete_mode": "single"})
        from memgo_cli.cli.commands.memory import cmd_delete

        backend = _get_backend(api_key, base_url)
        cmd_delete(
            backend,
            memory_id,
            dry_run=dry_run,
            force=force,
            delete_linked=delete_linked,
            output=output,
        )

    elif all_:
        _fire_telemetry("delete", {"delete_mode": "all"})
        from memgo_cli.cli.commands.memory import cmd_delete_all

        backend, config = _get_backend_and_config(api_key, base_url)
        ids = _resolve_ids(config, user_id=user_id, agent_id=agent_id, app_id=app_id, run_id=run_id)
        cmd_delete_all(backend, force=force, dry_run=dry_run, all_=project, **ids, output=output)

    else:  # 实体删除分支
        _fire_telemetry("delete", {"delete_mode": "entity"})
        from memgo_cli.cli.commands.entities import cmd_entities_delete

        backend = _get_backend(api_key, base_url)
        cmd_entities_delete(
            backend,
            user_id=user_id,
            agent_id=agent_id,
            app_id=app_id,
            run_id=run_id,
            force=force,
            dry_run=dry_run,
            output=output,
        )


def register(app: typer.Typer) -> None:
    """注册当前命令及其参数。"""
    app.command(
        rich_help_panel="Memory",
        help="Delete a memory, all memories, or an entity.\n\nExamples:\n  memgo delete abc-123-def-456\n  memgo delete abc-123 --dry-run\n  memgo delete --all -u alice --force\n  memgo delete --all --project --force\n  memgo delete --entity -u alice --force",
    )(delete)

File: cli/node/src/cli/registrations/memory/delete.ts
import type { Command } from 'commander';
import { printError } from '../../../output/branding.js';
import { exit } from '../../../runtime/exit.js';
import { setAgentMode } from '../../../runtime/state.js';

import { resolveIds } from '../../../application/entity-ids.js';
import { getBackendAndConfig, getBackendOnly } from '../../context.js';

/**
 * 注册 memory/delete 命令，保持参数和输出合同。
 * @param program - 本次装配的 Commander 根命令。
 */
export function registerMemoryDelete(program: Command): void {
  /**
   * 读取全局选项并同步本次调用的代理输出模式。
   * @returns 本次调用是否采用代理输出模式。
   */
  function checkAgentMode(): boolean {
    const value = !!(program.opts().json || program.opts().agent);
    setAgentMode(value);
    return value;
  }

  program
    .command('delete [memoryId]')
    .description('Delete a memory, all memories matching a scope, or an entity.')
    .option('--all', 'Delete all memories matching scope filters.', false)
    .option('--entity', 'Delete the entity itself and all its memories (cascade).', false)
    .option('--project', 'With --all: delete ALL memories project-wide.', false)
    .option('--dry-run', 'Show what would be deleted without deleting.', false)
    .option('--force', 'Skip confirmation.', false)
    .option('--delete-linked', 'Also delete memories linked to this memory.', false)
    .option('-u, --user-id <id>', 'Scope to user.')
    .option('--agent-id <id>', 'Scope to agent.')
    .option('--app-id <id>', 'Scope to app.')
    .option('--run-id <id>', 'Scope to run.')
    .option('-o, --output <format>', 'Output: text, json, quiet.', 'text')
    .option('--api-key <key>', 'Override API key.')
    .option('--base-url <url>', 'Override API base URL.')
    .addHelpText(
      'after',
      [
        '\nExamples:',
        '  $ memgo delete abc-123-def-456              # single memory',
        '  $ memgo delete --all -u alice --force        # all memories for user',
        '  $ memgo delete --all --project --force       # project-wide wipe',
        '  $ memgo delete --entity -u alice --force     # entity + all its memories',
      ].join('\n')
    )
    .action(async (memoryId, opts) => {
      const isAgent = checkAgentMode();
      const output = isAgent ? 'agent' : opts.output;
      // 校验互斥删除参数。
      if (memoryId && opts.all) {
        printError('Cannot combine <memoryId> with --all. Use one or the other.');
        exit(1);
      }
      if (memoryId && opts.entity) {
        printError('Cannot combine <memoryId> with --entity. Use one or the other.');
        exit(1);
      }
      if (opts.all && opts.entity) {
        printError('Cannot combine --all with --entity. Use one or the other.');
        exit(1);
      }
      if (!memoryId && !opts.all && !opts.entity) {
        printError(
          'Specify a memory ID, --all, or --entity.\n' +
            '  memgo delete <id>              Delete a single memory\n' +
            '  memgo delete --all [scope]     Delete all memories matching scope\n' +
            '  memgo delete --entity [scope]  Delete an entity and all its memories'
        );
        exit(1);
      }

      // 删除单条记忆。
      if (memoryId) {
        const { cmdDelete } = await import('../../../cli/commands/memory/delete.js');
        const backend = await getBackendOnly(opts.apiKey, opts.baseUrl);
        await cmdDelete(backend, memoryId, {
          output,
          dryRun: opts.dryRun,
          force: opts.force,
          deleteLinked: opts.deleteLinked,
        });
        return;
      }

      // 删除范围内全部记忆。
      if (opts.all) {
        const { cmdDeleteAll } = await import('../../../cli/commands/memory/delete-all.js');
        const { backend, config } = await getBackendAndConfig(opts.apiKey, opts.baseUrl);
        const ids = opts.project
          ? {
            userId: undefined,
            agentId: undefined,
            appId: undefined,
            runId: undefined,
          }
          : resolveIds(config, opts);
        await cmdDeleteAll(backend, {
          force: opts.force,
          dryRun: opts.dryRun,
          all: opts.project,
          ...ids,
          output,
        });
        return;
      }

      // 删除实体及其记忆。
      if (opts.entity) {
        const { cmdEntitiesDelete } = await import('../../commands/entities.js');
        const backend = await getBackendOnly(opts.apiKey, opts.baseUrl);
        await cmdEntitiesDelete(backend, { ...opts, output });
        return;
      }
    });
}
